src/v/raft/mutex_buffer.h

```diff
@@ -0,0 +1,123 @@
+#pragma once
+#include "raft/types.h"
+#include "ssx/future-util.h"
+#include "utils/mutex.h"
+
+#include <seastar/core/condition-variable.hh>
+#include <seastar/core/do_with.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/gate.hh>
+#include <seastar/core/loop.hh>
+#include <seastar/core/shared_ptr.hh>
+
+namespace raft::details {
+
+template<typename Request, typename Response>
+class mutex_buffer {
+public:
+    explicit mutex_buffer(mutex& m)
+      : _mutex(m) {}
+
+    ss::future<Response> enqueue(Request&& r) {
+        if (_gate.is_closed()) {
+            return ss::make_exception_future<Response>(
+              ss::gate_closed_exception{});
+        }
+
+        ss::promise<Response> p;
+        auto f = p.get_future();
+        _requests.push_back(std::move(r));
+        _responsens.push_back(std::move(p));
+        _enequeued.signal();
+        return f;
+    }
+    // clang-format off
+    template<typename Func>
+    CONCEPT(requires requires(Func f, Request req) {
+        { f(std::move(req)) } -> ss::future<Response>;
+    })
+    // clang-format on
+    void start(Func&& f);
+
+    ss::future<> stop() {
+        _enequeued.broken();
+        return _gate.close();
+    }
+
+private:
+    using response_t = std::vector<ss::promise<Response>>;
+    using request_t = std::vector<Request>;
+
+    template<typename Func>
+    ss::future<> flush(Func&& f);
+    template<typename Func>
+    ss::future<> do_flush(request_t, response_t, Func&& f);
+
+    request_t _requests;
+    response_t _responsens;
+    ss::condition_variable _enequeued;
```

Comment:
we might need a semaphore of say 256 units or smth like that, in addition to the condition variable.